scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,359 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Collection of functions for
+ - Addition and removal of exports
+ - Management of client permissions on export locations
+"""
+
+import errno
+import functools
+import io
+import json
+import logging
+import os
+import subprocess
+import sys
+import time
+
+try:
+    import ConfigParser as configparser
+except ImportError:
+    import configparser
+
+
+from scality_manila_utils import utils
+from scality_manila_utils.exceptions import (ClientExistsException,
+                                             ClientNotFoundException,
+                                             EnvironmentException,
+                                             ExportException,
+                                             ExportAlreadyExists,
+                                             ExportNotFoundException,
+                                             ExportHasGrantsException)
+
+log = logging.getLogger(__name__)
+if sys.version_info[0] == 2:
+    text_type = unicode
+else:
+    text_type = str
+
+
+# From http://prosseek.blogspot.fr/2012/10/
+# reading-ini-file-into-dictionary-in.html
+class MyIniParser(configparser.ConfigParser):
+    def as_dict(self):
+        d = dict(self._sections)
+        for k in d:
+            d[k] = dict(self._defaults, **d[k])
+            d[k].pop('__name__', None)
+        return d
+
+
+def _get_defined_exports():
+    """Retrieve all defined exports from the Samba registry."""
+
+    with utils.elevated_privileges():
+        process = subprocess.Popen(['net', 'conf', 'list'],
+                                   stdout=subprocess.PIPE)
+        # `subprocess.communicate` returns a byte stream
+        output = text_type(process.communicate()[0])
+
+    if process.returncode != 0:
+        raise EnvironmentError("Something went wrong while dumping the "
+                               "Samba registry: '{0:s}'".format(output))
+
+    config = MyIniParser()
+    output = text_type(output.replace('\t', ''))
+    config.readfp(io.StringIO(output))
+
+    return config.as_dict()
+
+
+def verify_environment(root_export):
+    """
+    Preliminary checks for installed binaries and running services.
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :raises:
+        :py:class:`scality_manila_utils.exceptions.EnvironmentException`
+        if the environment is not ready
+    """
+
+    if not utils.is_stored_on_sofs(root_export):
+        raise EnvironmentException("%s doesn't seem to be stored on a SOFS "
+                                   "filesystem" % root_export)
+
+    env_path = os.getenv('PATH').split(':')
+    for binary in ('net', 'sfused'):
+        utils.binary_check(binary, env_path)
+
+    for process in ('sfused', 'smbd'):
+        utils.process_check(process)
+
+    with open('/etc/samba/smb.conf') as f:
```

Comment:
`io.open` provides slightly more uniform behavior between python version 2 and 3. In the case above you would have a unicode str in both cases.
When using the builtin `open` on python2 you will get a regular bytestr, and on python 3 you will get a unicode str.